### PR TITLE
IGNITE-12568 MessageFactory refactoring

### PR DESCRIPTION
**Pull request by agura**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/IgniteMessageFactoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+        if (!old.isEmpty()) {
+            for (int i = 0; i < ARR_SIZE; i++) {
+                Supplier<Message> curr = msgSuppliers[i];
+
+                if (curr == null) {
+                    short directType = indexToDirectType(i);
+
+                    for (MessageFactory factory : old) {
+                        Message msg = factory.create(directType);
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/communication/IgniteMessageFactoryImpl.java]: Should you wrap this into try-catch?
Specification isn't very clear - it says that method returns null for unresolved direct type but at the same time this class (IgniteMessageFactoryImpl) throws exceptions for that same reason. We should probably unify this behavior and correct tests correspondingly.

Comment (agura) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/communication/IgniteMessageFactoryImpl.java]: I agree with you, but:
- at present we have several usages of `msgFactory.create(type)` where `null` is valid but treated in different way (e.g. `DirectByteBufferStreamImplV2#readMessage`);
- exception will thrown to higher level and in current implementation and architecture it is ok.

In the future (possibly during 3.0 development) we should return to this issue.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/IgniteMessageFactoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+     * @throws IgniteException If there are no any message factory for given {@code directType}.
+     */
+    @Override public @Nullable Message create(short directType) {
+        Supplier<Message> supplier = msgSuppliers[directTypeToIndex(directType)];
+
+        if (supplier == null)
+            supplier = CUSTOM.get(directType);
+
+        if (supplier == null)
+                throw new IgniteException("Invalid message type: " + directType);
```
[the remark this answers is not in the document]
Comment (agura) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/communication/IgniteMessageFactoryImpl.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 3]
```diff
@@ -103,7 +103,7 @@
     private final Object msgTypMapMux = new Object();
 
     /** Message type map. */
-    private volatile Map<Short, String> msgTypMap;
+    private volatile Map<Short, String> msgTypMape;
```
[the remark this answers is not in the document]
Comment (agura) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: I mean `msgTypeMap` :) Fixed.

File: modules/core/src/main/java/org/apache/ignite/plugin/extensions/communication/MessageFactoryProvider.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.plugin.extensions.communication;
+
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Provider of communication message fctories.
```
[the remark this answers is not in the document]
Comment (agura) [reply modules/core/src/main/java/org/apache/ignite/plugin/extensions/communication/MessageFactoryProvider.java]: Fixed.

**Comment by agura** [comment]
Merged to master branch.

**Pull request closed without merging**